### DISCUSSION TITLE
Noticeable cursor/input delay issue with specific Wacom Tablet driver

### DISCUSSION THREAD

**Discussion by RatinFX** [ppy/osu#28951] [Q&A]
Hey, I'm having **constant** (not for X amount of time or random) cursor/input delay issues only inside lazer, outside of the client it works as expected.
It's a noticeable difference compared to stable, it's also reproducible from what I've seen and tested with other people.

Setup:
- Tablet: **CTL/CTH480** + **Wacom 6.3.9w5 driver**, which is famous for being very responsible,
  - I've tried OpenTabletDriver but no combination of settings came close to the original,
- Monitor: 1080p with 144Hz, though it might not matter much here.

In-game and settings:
- Song select, Gameplay, Editor, no place is safe from this delay,
- Different Skins, toggling on/off every option made no difference,
- I also tried all possible combinations of Settings + Borderless / Full Screen + 4x / 8x / Basically unlimited fps, and they all resulted in the same cursor/input delay,
  - Windowed + VSync / 2x was performing noticeably worse so I didn't test them further.

Any ideas, recommendations, or more information to provide?

**Comment by Stoppedpuma** [comment]
> Wacom 6.3.9w5 driver

This driver is nearly 10 years old, if anything it's likely to be an issue with the fact that you're using such an outdated driver.

Can you try the usual steps of disabling windows ink and turning off or on "High Precision Input" to see if that makes a difference?

**Reply by RatinFX** [reply]
> Can you try the usual steps of disabling windows ink and turning off or on "High Precision Input" to see if that makes a difference?

Sadly any combination of these leaves me with a cursor that cannot be moved via the Pen.

It is indeed a very old driver, yet it's the only one to provide such high responsibility while working fine everywhere else outside of lazer.

From more testing - it may or may not be the case but - it feels like there's a built-in smoothing whenever my cursor enters the client's window.

**Reply by Stoppedpuma** [reply]
> From more testing - it may or may not be the case but - it feels like there's a built-in smoothing whenever my cursor enters the client's window.

You might just be noticing the difference between hardware cursor and drawn cursor. Can you see if the latency feels the same with capping both osu! clients to the same framerates? You can do this by setting these lines in `osu!.username.cfg` for stable:

`FrameSync = Custom`
`CustomFrameLimit = 1000` (Change 1000 to something lower if you can't achieve it on both clients. e.g 576 for 4x)

**Reply by RatinFX** [reply]
> might just be noticing the difference between hardware cursor and drawn cursor

Right, thanks for clarifying, I kind of assumed it was something like that.

> Can you see if the latency feels the same with capping both osu! clients to the same framerates?

Yeah sure, input-feel-wise stable comes out on top in all scenarios, though on 2x it introduced random stutters while lazer stayed the same but that's a bit unrelated.

I start to think that part of this might be a PC issue too as I cannot get too many frames on lazer:
- stable: Unlimited is around 0.8~1.0ms which could be a bit above 1k fps (+8x),
- lazer: regardless of the current FPS limit, generally hovers between 330-400fps (sometimes dipping below 300 on miss, etc.),
  - with the detailed FPS display it's around 300-330.

**Reply by Stoppedpuma** [reply]
> 330-400fps

This would probably explain why you feel more input lag, can I have some more information about your system to determine if this is an issue on your end or just lazer not being as optimised?

**Reply by RatinFX** [reply]
Sure, anything other than these?
- 16GB RAM
- GTX 1060 3GB
- AMD Ryzen 5 1600
- (+ Windows 11, same thing happened on 10 as well)

**Reply by Stoppedpuma** [reply]
I'm going to say this is a mixture of both as an old machine of mine with a 1050ti and Ryzen 5 5600x is getting ~600-800 fps, one big difference here though is this machine is running Gentoo instead of Windows so take that as you will. Have you tried different renderers?

**Reply by RatinFX** [reply]
Yup, the same thing is present with all of them with generally the same amount of FPS, except the game sometimes breaks with Vulkan.

**Reply by Stoppedpuma** [reply]
> the game sometimes breaks with Vulkan.

This is to be expected as it's still experimental.

I recommend trying the usual things such as updating drivers, making sure you don't have heavy processes running (Wallpaper engine seems to cause a lot of issues), making sure you have your power profile set to high performance, ensuring bios settings related to performance are enabled, etc. You can find some other recommendations [here](https://osu.ppy.sh/wiki/en/Performance_troubleshooting)! 

If you're still having problems I would say this is likely either something you have running in the background causing issues or your windows install as they tend to just cause more issues the older they are in my experience. I don't believe you should be receiving this performance as I'm able to double or sometimes even nearly triple your performance numbers with mostly lower end hardware than what's present in your system.

**Reply by RatinFX** [reply]
I generally don't have many things running in the background for that purpose except for Discord. I also tested it without anything else running, and there wasn't much difference.

I went back to re-check everything and I noticed the game was set to Single Threaded for some reason 😓
Changing it to Multi Threaded the game runs between 700-900fps during gameplay on Full Screen.

However, the issue is still present although it's less noticeable sometimes.
Weirdly enough, it felt quite a bit more responsive on Borderless with 500-600fps so there's that lol.

**Reply by Stoppedpuma** [reply]
> it felt quite a bit more responsive on Borderless with 500-600fps so there's that lol.

Weird, no clue why this would be the case? I'm going to say this is more than likely the original guess of it being an issue with the driver. I don't think much more can be said and this will likely remain unfixed as this is a driver which is nearly 10 years old. My advice would be to experiment with OpenTabletDriver some more to try to find something which feels comfortable as a driver this old is likely to have some serious security issues.
